build_runner_core/lib/src/asset/writer.dart

```diff
@@ -3,12 +3,123 @@
 // BSD-style license that can be found in the LICENSE file.
 
 import 'dart:async';
+import 'dart:convert';
 
 import 'package:build/build.dart';
+import 'package:glob/glob.dart';
+
+import 'reader.dart';
 
 @Deprecated('No longer used')
 typedef OnDelete = void Function(AssetId id);
 
 abstract class RunnerAssetWriter implements AssetWriter {
   Future delete(AssetId id);
+
+  /// Potentially runs pending work that needs to happen just before the build
+  /// completes, after all other interactions with this [AssetWriter] have
+  /// already been completed.
+  FutureOr<void> onBuildComplete() {}
+}
+
+/// A [RunnerAssetWriter] implementation that performs all writes at once in
+/// [onBuildComplete].
+///
+/// This write behavior may reduce load on external tools doing work on each
+/// file system change (like the analysis server).
+class DelayedAssetWriter implements RunnerAssetWriter {
+  final RunnerAssetWriter _delegate;
+
+  final Map<AssetId, List<int>?> overlay = {};
```

Comment:
I use `null` to express the state that an asset has been deleted (via `delete()`)  without the delayed writer being flushed (via `onBuildComplete`).

I've added a comment explaining the structure. Should I store information about deleted files separately to avoid the nullable type instead?

Reply:
That seems fine to me, it also handles cleanly the case where a file is deleted at the start of a build, but then re-written later.